ludwig/data/cache/util.py

```diff
@@ -1,18 +1,26 @@
 import ludwig
-from ludwig.constants import DEFAULTS, INPUT_FEATURES, NAME, OUTPUT_FEATURES, PREPROCESSING, TYPE
+from ludwig.constants import DEFAULTS, INPUT_FEATURES, OUTPUT_FEATURES, PREPROCESSING, PROC_COLUMN, TYPE
 from ludwig.data.cache.types import CacheableDataset
 from ludwig.types import ModelConfigDict
 from ludwig.utils.data_utils import hash_dict
 
 
 def calculate_checksum(original_dataset: CacheableDataset, config: ModelConfigDict):
+    """Calculates a checksum for a dataset and model config.
+
+    The checksum is used to determine if the dataset and model config have changed since the last time the model was
+    trained. If either has changed, a different checksum will be produced which will lead to a cache miss and force
+    preprocessing to be performed again.
+    """
     features = config.get(INPUT_FEATURES, []) + config.get(OUTPUT_FEATURES, []) + config.get("features", [])
     info = {
         "ludwig_version": ludwig.globals.LUDWIG_VERSION,
         "dataset_checksum": original_dataset.checksum,
         "global_preprocessing": config.get(PREPROCESSING, {}),
         "global_defaults": config.get(DEFAULTS, {}),
-        "feature_names": [feature[NAME] for feature in features],
+        # PROC_COLUMN contains both the feature name and the feature hash that is computed
+        # based on each feature's preprocessing parameters and the feature's type.
+        "feature_proc_columns": {feature[PROC_COLUMN] for feature in features},
```

Comment:
This is such a fundamental change because PROC_COLUMN is a combination of `feature_name`_`hash of dict of type + preprocessing parameters`

So a change to either the feature's name, type or preprocessing parameters will cause a change in name to both proc col and well as compute checksum. 

The second thing to note is that we're more or less guaranteed different preprocessing hashes for two different features anyway since the preprocessing for category vs binary (internal representation after all internal parameters are merged) are completely different!